src/lib/optimizer/strategy/predicate_pushdown_rule.cpp

```diff
@@ -54,17 +69,24 @@ bool PredicatePushdownRule::apply_to(const std::shared_ptr<AbstractLQPNode>& nod
     return true;
 
   } else if (input->type == LQPNodeType::Sort) {
-    // push always down if other node is a sort node
-    const auto sort_node = std::dynamic_pointer_cast<SortNode>(input);
-
-    lqp_remove_node(node);
-    const auto previous_left_input = sort_node->left_input();
-
-    sort_node->set_left_input(node);
-    node->set_left_input(previous_left_input);
-
+    // always push down if other node is a sort node
+    push_down(node, input);
+    return true;
+  } else if (input->type == LQPNodeType::Projection) {
+    // push below projection if the projection does not generate the column(s) that we are scanning on
+    const auto projection_input = input->left_input();
```

Comment:
This whole block can probably be replaced by
`expression_evaluable_on_lqp(predicate_node->predicate, input->left_input())`

...this will probably fit you "free functions are evil" narrative... ;)

Reply:
You guessed right. This is a nice function, but how are people supposed to know about it? :(

Reply:
(This particular method could probably be part of `AbstractExpression`, it feels pretty native to the class.)

But in general, though not properly satisfying, I know: By knowing that in Hyrise the are `*_utils.hpp` files that define useful functions for their component. 
And when working in this component, you should familiarize yourself with both the classes and functions in this component, not only with the classes.

Reply:
Looking at the test fails, it seems as if `expression_evaluable_on_lqp` is not well-suited for this. For example, for `SELECT * FROM id_int_int_int_100 WHERE a IN (SELECT b FROM mixed)` it claims that `a IN (SELECT b FROM mixed)` is evaluable even before the `ProjectionNode`.

It should check for things like `IN`, where the boolean result is only created by the Projection. There are probably other cases that I can't think of right now, too.

Also, maybe `expression_available_in_lqp` would be a better name.

Reply:
Sorry, my mistake: `expression_evaluable_on_lqp` checks whether an expression can be *evaluated by the ExpressionEvaluator* on top of a specified LQP. It does NOT return whether it can be *evaluated by a TableScan* on top of a specified node. Get what I mean?

Instead, `OperatorScanPredicate::from_expression(predicate_node->predicate, input->left_input) != std::nullopt` will tell you whether an expression can be turned into a TableScan at a specified position in the LQP.